gapic-common/lib/gapic/rest/paged_enumerable.rb

```diff
@@ -0,0 +1,226 @@
+        end
+      end
+
+      ##
+      # Iterate over the pages.
+      #
+      # @yield [Page] Gives the pages in the stream.
+      #
+      # @raise if it's not started yet.
+      # @return [Enumerable, Nil]
+      def each_page
+        return enum_for :each_page unless block_given?
+
+        loop do
+          break if @page.nil?
+          yield @page
+          next_page!
+        end
+      end
+
+      ##
+      # True if it has the next page.
+      #
+      # @return [Boolean]
+      #
+      def next_page?
+        @page.next_page_token?
+      end
+
+      ##
+      # Update the response in the current page.
+      #
+      # @return [Page] the new page object.
+      #
+      def next_page!
+        unless next_page?
+          @page = nil
+          return @page
+        end
+
+        next_request = @request.dup
+        next_request.page_token = @page.next_page_token
+
+        @response = @service_stub.send @method_name, next_request, @options
+        @page = Page.new @response, @resource_field_name, format_resource: @format_resource
+      end
+      alias next_page next_page!
+
+      ##
+      # The page token to be used for the next RPC call.
+      #
+      # @return [String]
+      #
+      def next_page_token
+        @page.next_page_token
+      end
+
+      ##
+      # The current response object, for the current page.
+      #
+      # @return [Object]
+      #
+      def response
+        @page.response
+      end
+
+      ##
+      # A class to represent a page in a PagedEnumerable. This also implements Enumerable, so it can iterate over the
+      # resource elements.
+      #
+      # @attribute [r] response
+      #   @return [Object] the response object for the page.
+      class Page
+        include Enumerable
+        attr_reader :response
+
+        ##
+        # @private
+        # @param response [Object] The response object for the page.
+        # @param resource_field [String] The name of the field in response which holds the resources.
+        # @param format_resource [Proc] A Proc object to format the resource object.
+        # The Proc should accept response as an argument, and return a formatted resource object. Optional.
+        #
+        def initialize response, resource_field, format_resource: nil
+          @response = response
+          @resource_field = resource_field
+          @format_resource = format_resource
+        end
+
+        ##
+        # Iterate over the resources.
+        #
+        # @yield [Object] Gives the resource objects in the page.
+        #
+        def each
+          return enum_for :each unless block_given?
+
+          return if resources.nil?
+
+          # We trust that the field exists and is an Enumerable
+          resources.each do |resource|
+            resource = @format_resource.call resource if @format_resource
+            yield resource
+          end
+        end
+
+        ##
+        # The page token to be used for the next RPC call.
```

Comment:
We should probably document what happens if there is no next page. (I think it returns the empty string?)

Reply:
:+1:

Reply:
document? (e.g. `The page token to be used for the next RPC call, or the empty string if there is no next page.`)

Reply:
:+1:  :+1: